Android Appium test tweaks

filmaj opened the pull request:
<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist
is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected

Android

### What does this PR do?

- updated comments on how to run the tests.
- extra comments around functionality at certain points in the automation.
- stub of a resolution checker on test startup - still need to figure out acceptable values.
- moved session shutdown to an `afterAll` clause. 
- changed resolution determiner from using webview-based values to using the native windows dimensions. this helps as the webview values may be scaled down intentionally by manufacturers (via changing `window.devicePixelRatio`, for example). furthermore, since the screen dimensions are used purely for native UI automation using direct coordinates, better to use the dimensions reported by the native context rather than the web context.
- removed duplicate element calls.

### What testing has been done on this change?

Tested on local appium 1.6.3 environment on Android 4.4 and 5.1 stock emulators.

### Checklist
- [x] [Reported an issue](http://cordova.apache.org/contribute/issues.html) in the JIRA database
- [x] Commit message follows the format: "CB-3232: (android) Fix bug with resolving file paths", where CB-xxxx is the JIRA ID & "android" is the platform affected.
- [ ] Added automated test coverage as appropriate for this change.

cordova-qa commented:
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/244/commits/f4616d88cd1b87433f0f8b98a1d228253be461aa)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/110//PLATFORM=android/artifact/) |

filmaj commented:
Oh! I totally forgot to ping @alsorokin for a review :)

If you find the time, please review good sir!

filmaj commented on appium-tests/android/android.spec.js:
```diff
@@ -21,10 +21,13 @@
 // it is not necessary to do a full CI setup to run these tests
-// just run "node cordova-medic/medic/medic.js appium --platform android --plugins cordova-plugin-camera"
+// for android 5.1, run:
+//      node cordova-paramedic/main.js --platform android --plugin cordova-plugin-camera --skipMainTests --saucePlatformVersion "5.1" --target api22
```
So, while that may be how paramedic right now was designed, that's actually counter-productive when working with appium. Appium will actually select the correct emulator (and even start it for you) based on the `platformName`, `platformVersion` and `deviceName` capabilities you provide. These capabilities apply regardless if you are running the test locally or on Sauce Labs.

I think, moving forward, for a paramedic refactor I am gathering use cases for / noting issues to address, we would want to expose the appium/selenium capabilities more completely. This would also reduce the difference between local test executions and executions on Sauce Labs, by allowing Appium to worry about managing the emulator, rather than managing it in paramedic.

I will change the instructions back, but just a heads up that I think we should change this behaviour in the future.

alsorokin replied:
I still think that the user should know better which emulator he wants to use in the test run. For example, one of the emulator may not have camera enabled. I doubt that Appium would know that when selecting "best" emulator.
Also please don't forget that `--target` arg specifies not only emulator for Appium tests, but also for our regular "mobilespec" ones. It just seems natural to me that the Appium should use already opened and ready-for-action emulator instead of choosing another one.
I'm not against any changes in paramedic, just letting you know how it works right now. 😉

filmaj commented on appium-tests/android/android.spec.js:
```diff
@@ -373,7 +384,6 @@ describe('Camera tests Android.', function () {
                         return driver
                             .waitForElementByXPath('//android.widget.TextView[@text="Gallery"]', 20000)
                             .elementByXPath('//android.widget.TextView[@text="Gallery"]')
-                            .elementByXPath('//android.widget.TextView[@text="Gallery"]')
```
Do you have any details around the emulator bug on Windows? Ideally:
 - any Appium issues filed that sound similar?
 - any Android issues filed relevant?
 - Windows version?
 - Appium version?
 - Android emulator version?

I'm also up for looking into it more closely.

alsorokin replied:
I've tried to find any similar issues but with no luck.
Windows: Windows 10 Enterprise, Version 1607, Build 14393.577
Appium: 1.5.0 (maybe even earlier) to 1.6.3
Android SDK Tools (i.e. emulator version) 25.2.2

filmaj commented:
@alsorokin thanks for the feedback, I've addressed your points. Let me know if you have any other concerns or if you are good with merging it.

cordova-qa commented:
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/244/commits/4e26e2a04ae256e4305419a9d469771dd4e0f8c8)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/113//PLATFORM=android/artifact/) |

alsorokin commented:
LGTM 👍

alsorokin commented:
Nit: it would be nice if first two checklist items would be fulfilled.

filmaj commented:
Filed [CB-12312](https://issues.apache.org/jira/browse/CB-12312) for this one.

filmaj commented:
Rebased and squashed commit and updated message to have the same format as you pointed out @alsorokin. Thanks for the review! I'll merge upon tests passing.

cordova-qa commented:
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/244/commits/05594c46468576da846b31b03e624d72f7c9c476)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=windows-10-store/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/115//PLATFORM=android/artifact/) |
